storage/provider/openstack_test.go

```diff
@@ -0,0 +1,356 @@
+package provider
+
+import (
+	"bytes"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/juju/juju/instance"
+	"github.com/juju/juju/storage"
+
+	"github.com/juju/names"
+	jc "github.com/juju/testing/checkers"
+
+	gc "gopkg.in/check.v1"
+	"gopkg.in/goose.v1/cinder"
+	"gopkg.in/goose.v1/client"
+	goosehttp "gopkg.in/goose.v1/http"
+	"gopkg.in/goose.v1/nova"
+)
+
+const (
+	mockVolId           = "0"
+	mockVolSize  uint64 = 1024 * 2
+	mockVolName         = "mock-volume-name"
+	mockServerId        = "mock-server-id"
+	mockVolJson         = `{"volume":{"id": "` + mockVolId + `", "size":1,"name":"` + mockVolName + `"}}`
+)
+
+var _ = gc.Suite(&openstackSuite{})
+
+type openstackSuite struct{}
+
+func (s *openstackSuite) TestVolumeSource(c *gc.C) {
+
+	p := &openstackProvider{mockClientFactoryFn(nil)}
+
+	// Check that we're validating the config passed in.
+	cfg, err := storage.NewConfig("openstack", OpenstackProviderType, map[string]interface{}{})
+	c.Assert(err, jc.ErrorIsNil)
+	_, err = p.VolumeSource(nil, cfg)
+	c.Check(err, gc.ErrorMatches, "requisite configuration was not set: auth-url not assigned")
+
+	cfg, err = NewOpenstackStorageConfig("", "", "", "", "")
+	c.Assert(err, jc.ErrorIsNil)
+
+	volSource, err := p.VolumeSource(nil, cfg)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Check(volSource, gc.NotNil)
+}
+
+func (s *openstackSuite) TestAttachVolumes(c *gc.C) {
+
+	numCalls := 0
+	mockAdapter := &mockAdapter{
```

Comment:
Just as the AWS test server was extended to listen and respond to EBS requests, I'm in 2 minds as to whether we should be doing that here too. All the other openstack tests use the openstack test server. I'd like to discuss this further.